[c10d] Change collective to take in a list of tensors so it work fully for all collectives

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #135049

We found that currently, we only pass one input and output tensor to the function `collective`, and this causes NaNCheck, work numel stats and FR input/output sizes not accurate for all-to-all, scatter and reduce. So we want to let the collective take in a list of tensors to ensure it works for all collectives inside PGNCCL.

This partially revert what we did in https://github.com/pytorch/pytorch/pull/119421, and down the road we will have another round of cleanup on the collective to make it cleaner. For now, at least for the sake of correctness, we changed it back.

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @wz337 @wconstab @d4l3k @c-p-i-o

fduwjj committed:
[c10d] Change collective to take in a list of tensors so it work fully for all collectives

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Change collective to take in a list of tensors so it work fully for all collectives"


We found that currently, we only pass one input and output tensor to the function `collective`, and this causes NaNCheck, work numel stats and FR input/output sizes not accurate for all-to-all, scatter and reduce. So we want to let the collective take in a list of tensors to ensure it works for all collectives inside PGNCCL.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Change collective to take in a list of tensors so it work fully for all collectives"


We found that currently, we only pass one input and output tensor to the function `collective`, and this causes NaNCheck, work numel stats and FR input/output sizes not accurate for all-to-all, scatter and reduce. So we want to let the collective take in a list of tensors to ensure it works for all collectives inside PGNCCL.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented:
General comment:
how about using a redirect instead of altering every call site of `collective()`? 

A redirect can look like this:
```
collective(input, output) {
  inputs = vector{input};
  outputs = vector{output};
  collective(inputs, outputs);
}
```

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2721,25 +2720,25 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::collective(
       c10::cuda::CUDACachingAllocator::recordStream(
-          input.storage().data_ptr(), ncclStream);
+          inputs[0].storage().data_ptr(), ncclStream);
     } else {
       // for sparse input case record streams on both index and value
       // tensors
       c10::cuda::CUDACachingAllocator::recordStream(
-          input.values().storage().data_ptr(), ncclStream);
+          inputs[0].values().storage().data_ptr(), ncclStream);
       c10::cuda::CUDACachingAllocator::recordStream(
-          input.indices().storage().data_ptr(), ncclStream);
+          inputs[0].indices().storage().data_ptr(), ncclStream);
     }
   }
 #ifndef NCCL_HAS_COMM_NONBLOCKING
   C10D_NCCL_CHECK(
-      fn(input, output, comm, ncclStream),
+      fn(inputs[0], outputs[0], comm, ncclStream),
       ncclComm->getNcclCommFailureReason());
 #else
   C10D_NCCL_CHECK_TIMEOUT(
-      fn(input, output, comm, ncclStream),
+      fn(inputs[0], outputs[0], comm, ncclStream),
       comm,
       ncclComm->getNcclCommFailureReason());
 #endif
```
I think this style is still awkward. 

For record stream, we can probably iterate over every input to make it look good.

For fn(...), I don't know what style makes sense.. The definition of lambda itself has a mismatch with what input(s) are actually used -- the actual inputs are recorded rather than passed in..

fduwjj replied:
yep... I propose to keep the fn as it is for now.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2691,16 +2688,18 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::collective(
     work->stashed_for_allocator_safety_ =
         std::make_shared<std::vector<at::Tensor>>();
-    work->stashed_for_allocator_safety_->push_back(input);
+    work->stashed_for_allocator_safety_->push_back(inputs[0]);
   }
```
Maybe check if we need to push the entire inputs into `stashed_for_allocator_safety_`

fduwjj replied:
For alltoall, we do this in preprocess func.. lol

hippocookie commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2691,16 +2688,18 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::collective(
       std::make_shared<std::vector<at::Tensor>>(std::move(outputs));
 
-  if (avoidRecordStreams) {
+  if (avoidRecordStreams && inputs.size() == 0) {
```
Hi, I think here is check `inputs.size() > 0`, since I saw get `inputs[0]` in below, is that right?

fduwjj committed:
Update on "[c10d] Change collective to take in a list of tensors so it work fully for all collectives"


We found that currently, we only pass one input and output tensor to the function `collective`, and this causes NaNCheck, work numel stats and FR input/output sizes not accurate for all-to-all, scatter and reduce. So we want to let the collective take in a list of tensors to ensure it works for all collectives inside PGNCCL.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Change collective to take in a list of tensors so it work fully for all collectives"


We found that currently, we only pass one input and output tensor to the function `collective`, and this causes NaNCheck, work numel stats and FR input/output sizes not accurate for all-to-all, scatter and reduce. So we want to let the collective take in a list of tensors to ensure it works for all collectives inside PGNCCL.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -3224,6 +3231,33 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::pointToPoint(
+      inputs,
+      outputs,
+      fn,
+      [](at::cuda::CUDAStream&,
+         c10::intrusive_ptr<ProcessGroupNCCL::WorkNCCL>& work) {},
+      [](at::cuda::CUDAStream&,
+         c10::intrusive_ptr<ProcessGroupNCCL::WorkNCCL>& work) {},
```
I think we should pass on the `pre` and `post` as is.

fduwjj replied:
They call a short version of function which inside, it add an empty pre and post?

fduwjj replied:
Opps, I just do copy, paste, my bad.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2691,16 +2688,18 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::collective(
     work->stashed_for_allocator_safety_ =
         std::make_shared<std::vector<at::Tensor>>();
-    work->stashed_for_allocator_safety_->push_back(input);
+    work->stashed_for_allocator_safety_->push_back(inputs[0]);
   }
```
Stashing all `inputs`?

fduwjj replied:
Probably no, because in all-to-all we did that in preprocess.

kwen2501 replied:
I think it is better to think from THIS function's perspective (to make it standalone).

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2721,25 +2720,27 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::collective(
   }
 #ifndef NCCL_HAS_COMM_NONBLOCKING
   C10D_NCCL_CHECK(
-      fn(input, output, comm, ncclStream),
+      fn(inputs[0], outputs[0], comm, ncclStream),
       ncclComm->getNcclCommFailureReason());
 #else
   C10D_NCCL_CHECK_TIMEOUT(
-      fn(input, output, comm, ncclStream),
+      fn(inputs[0], outputs[0], comm, ncclStream),
       comm,
       ncclComm->getNcclCommFailureReason());
 #endif
```
nit: would be nice to add some explanation on why `[0]` *can* be used here (because lambda's records arguments at site of definition). Plus a TODO to clean this in future (by either entirely removing lambda's or removing input and output from lambda's signature).

fduwjj replied:
Sure. I can add that.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -4421,9 +4454,10 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::gather(
   return collective(
-      inputTensor,
-      outputs[0], // just to fit the collective interface
+      inputTensors,
```
nit: I think it would be clearer to create a vector to "just to fit the collective interface" than directly reusing `inputTensors` here.
`inputTensors` carries the notion of multiple devices, but that's not what the `collective` API means here.

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -4514,8 +4552,8 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::scatter(
   return collective(
-      outputTensor,
-      inputs[0], // just to fit the collective interface
+      outputTensors,
```
nit: same here wrt `outputTensors`

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -757,8 +757,8 @@ class TORCH_API ProcessGroupNCCL : public Backend {
   c10::intrusive_ptr<Work> collective(
-      at::Tensor& input,
-      at::Tensor& output,
+      at::Tensor& inputs,
+      at::Tensor& outputs,
       Fn fn,
```
nit: why this name change?

fduwjj replied:
opps

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -767,8 +767,20 @@ class TORCH_API ProcessGroupNCCL : public Backend {
   c10::intrusive_ptr<Work> collective(
-      at::Tensor& input,
-      at::Tensor& output,
+      at::Tensor& inputs,
+      at::Tensor& outputs,
```
nit: why this name change?

fduwjj committed:
Update on "[c10d] Change collective to take in a list of tensors so it work fully for all collectives"


We found that currently, we only pass one input and output tensor to the function `collective`, and this causes NaNCheck, work numel stats and FR input/output sizes not accurate for all-to-all, scatter and reduce. So we want to let the collective take in a list of tensors to ensure it works for all collectives inside PGNCCL.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Change collective to take in a list of tensors so it work fully for all collectives"


We found that currently, we only pass one input and output tensor to the function `collective`, and this causes NaNCheck, work numel stats and FR input/output sizes not accurate for all-to-all, scatter and reduce. So we want to let the collective take in a list of tensors to ensure it works for all collectives inside PGNCCL.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 reviewed (approved):
Thanks for the changes. LGTM

wconstab commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -2692,13 +2689,17 @@ c10::intrusive_ptr<Work> ProcessGroupNCCL::collective(
   if (nanCheck) {
-    checkForNan(input, ncclStream);
+    for (const auto& input : inputs) {
+      checkForNan(input, ncclStream);
```
hold on a sec- are we re-introducing a NaN-check bug here, where some 'empty' inputs could be Nan-checked?

fduwjj replied:
no, for this case if it is for the inputs of all-to-all. What is our fix to the empty inputs? To just use inputs[0]?

kwen2501 replied:
@wconstab what do you mean by empty inputs? like dummy inputs in non-root ranks?

fduwjj committed:
Update on "[c10d] Change collective to take in a list of tensors so it work fully for all collectives"


We found that currently, we only pass one input and output tensor to the function `collective`, and this causes NaNCheck, work numel stats and FR input/output sizes not accurate for all-to-all, scatter and reduce. So we want to let the collective take in a list of tensors to ensure it works for all collectives inside PGNCCL.

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

fduwjj commented:
@pytorchbot merge
